File: backend/src/modules/globalTemplateLibrary/globalTemplateLibrary.service.js
const prisma = require('../../lib/prisma');
const { HttpError } = require('../../lib/httpError');
const { getGlobalStarterTemplateSeedRows } = require('../../lib/legionStarterEquipmentTemplates');

/** Serialize concurrent first-load seeding (Import modal fetches equipment + graphic lists in parallel). */
let ensureEquipmentStartersInFlight = null;

/** One-time style cleanup: legacy rows were named "Legion AHU", etc. */
async function stripLegacyLegionPrefixFromGlobalEquipmentNames() {
  await prisma.$executeRaw`
    UPDATE "GlobalEquipmentTemplate"
    SET name = TRIM(BOTH ' ' FROM REPLACE(name, 'Legion ', ''))
    WHERE name LIKE 'Legion %'
  `;
}

/**
 * If GlobalEquipmentTemplate has no rows, upsert starter templates (same data as prisma seed).
 * Avoids empty library when db seed was never run.
 */
async function ensureGlobalEquipmentStartersInDb() {
  if (ensureEquipmentStartersInFlight) {
    return ensureEquipmentStartersInFlight;
  }
  ensureEquipmentStartersInFlight = (async () => {
    try {
      const count = await prisma.globalEquipmentTemplate.count();
      if (count > 0) return;
      const rows = getGlobalStarterTemplateSeedRows();
      for (const r of rows) {
        await prisma.globalEquipmentTemplate.upsert({
          where: { id: r.id },
          update: {
            name: r.name,
            equipmentType: r.equipmentType,
            description: r.description,
            defaultGraphicName: r.defaultGraphicName,
            pointsJson: r.pointsJson,
            status: r.status,
          },
          create: {
            id: r.id,
            name: r.name,
            equipmentType: r.equipmentType,
            description: r.description,
            defaultGraphicName: r.defaultGraphicName,
            pointsJson: r.pointsJson,
            status: r.status,
          },
        });
      }
    } finally {
      ensureEquipmentStartersInFlight = null;
    }
  })();
  return ensureEquipmentStartersInFlight;
}

function normalizePointsJson(raw) {
  if (raw == null) return [];
  if (Array.isArray(raw)) return raw;
  if (typeof raw === 'string') {
    try {
      const p = JSON.parse(raw);
      return Array.isArray(p) ? p : [];
    } catch {
      return [];
    }
  }
  return [];
}

function countBindingsFromGraphicState(state) {
  if (!state || typeof state !== 'object') return 0;
  const objects = state.objects;
  if (!Array.isArray(objects)) return 0;
  return objects.filter((o) => o && o.pointBinding && String(o.pointBinding).trim()).length;
}

function equipmentListRowFromRecord(r) {
  const points = normalizePointsJson(r.pointsJson);
  return {
    id: r.id,
    name: r.name,
    equipmentType: r.equipmentType,
    pointCount: points.length,
    defaultGraphicName: r.defaultGraphicName ?? null,
  };
}

async function listEquipmentTemplates() {
  await stripLegacyLegionPrefixFromGlobalEquipmentNames();
  await ensureGlobalEquipmentStartersInDb();
  const rows = await prisma.globalEquipmentTemplate.findMany({
    where: { status: 'ACTIVE' },
    orderBy: [{ name: 'asc' }],
  });
  return rows.map((r) => equipmentListRowFromRecord(r));
}

async function updateEquipmentTemplateName(id, body) {
  const name = String(body?.name || '').trim();
  if (!name) throw new HttpError(400, 'name is required');
  let r;
  try {
    r = await prisma.globalEquipmentTemplate.update({
      where: { id },
      data: { name },
    });
  } catch (e) {
    if (e.code === 'P2025') throw new HttpError(404, 'Global equipment template not found');
    throw e;
  }
  return equipmentListRowFromRecord(r);
}

async function deleteEquipmentTemplate(id) {
  try {
    await prisma.globalEquipmentTemplate.delete({ where: { id } });
  } catch (e) {
    if (e.code === 'P2025') throw new HttpError(404, 'Global equipment template not found');
    throw e;
  }
}

async function getEquipmentTemplateById(id) {
  await stripLegacyLegionPrefixFromGlobalEquipmentNames();
  await ensureGlobalEquipmentStartersInDb();
  const r = await prisma.globalEquipmentTemplate.findFirst({
    where: { id, status: 'ACTIVE' },
  });
  if (!r) throw new HttpError(404, 'Global equipment template not found');
  const points = normalizePointsJson(r.pointsJson);
  return {
    id: r.id,
    name: r.name,
    equipmentType: r.equipmentType,
    description: r.description ?? '',
    defaultGraphicName: r.defaultGraphicName ?? null,
    pointCount: points.length,
    points,
  };
}

async function createEquipmentTemplateFromSitePayload(body) {
  if (!body || typeof body !== 'object') {
    throw new HttpError(400, 'Invalid body');
  }
  const name = String(body.name || '').trim();
  if (!name) throw new HttpError(400, 'name is required');
  const equipmentType = String(body.equipmentType || 'CUSTOM').trim() || 'CUSTOM';
  const description = String(body.description ?? '').trim();
  const defaultGraphicName =
    body.defaultGraphicName != null && String(body.defaultGraphicName).trim()
      ? String(body.defaultGraphicName).trim()
      : body.defaultGraphic != null && String(body.defaultGraphic).trim()
        ? String(body.defaultGraphic).trim()
        : null;
  const points = normalizePointsJson(body.points);

  const created = await prisma.globalEquipmentTemplate.create({
    data: {
      name,
      equipmentType,
      description,
      defaultGraphicName,
      pointsJson: points,
    },
  });

  return getEquipmentTemplateById(created.id);
}

function graphicListRowFromRecord(r) {
  return {
    id: r.id,
    name: r.name,
    appliesToEquipmentType: r.appliesToEquipmentType,
    boundPointCount: r.boundPointCount ?? 0,
    globalEquipmentTemplateId: r.globalEquipmentTemplateId ?? null,
    equipmentTemplateName: r.equipmentTemplateName ?? null,
  };
}

async function updateGraphicTemplateName(id, body) {
  const name = String(body?.name || '').trim();
  if (!name) throw new HttpError(400, 'name is required');
  let r;
  try {
    r = await prisma.globalGraphicTemplate.update({
      where: { id },
      data: { name },
    });
  } catch (e) {
    if (e.code === 'P2025') throw new HttpError(404, 'Global graphic template not found');
    throw e;
  }
  return graphicListRowFromRecord(r);
}

async function deleteGraphicTemplate(id) {
  try {
    await prisma.globalGraphicTemplate.delete({ where: { id } });
  } catch (e) {
    if (e.code === 'P2025') throw new HttpError(404, 'Global graphic template not found');
    throw e;
  }
}

async function listGraphicTemplates() {
  const rows = await prisma.globalGraphicTemplate.findMany({
    where: { status: 'ACTIVE' },
    orderBy: [{ name: 'asc' }],
  });
  return rows.map((r) => graphicListRowFromRecord(r));
}

async function getGraphicTemplateById(id) {
  const r = await prisma.globalGraphicTemplate.findFirst({
    where: { id, status: 'ACTIVE' },
  });
  if (!r) throw new HttpError(404, 'Global graphic template not found');
  const graphicEditorState =
    r.graphicEditorStateJson != null && typeof r.graphicEditorStateJson === 'object'
      ? r.graphicEditorStateJson
      : null;
  const boundFromState = countBindingsFromGraphicState(graphicEditorState);
  return {
    id: r.id,
    name: r.name,
    appliesToEquipmentType: r.appliesToEquipmentType,
    globalEquipmentTemplateId: r.globalEquipmentTemplateId ?? null,
    equipmentTemplateName: r.equipmentTemplateName ?? null,
    boundPointCount: Math.max(r.boundPointCount ?? 0, boundFromState),
    graphicEditorState,
  };
}

/**
 * @param {object} body - site graphic template row
 * @param {Array<{ id: string, name: string, equipmentType: string }>} [equipmentTemplates] - site equipment templates to resolve appliesTo / equipmentTemplateId
 */
async function createGraphicTemplateFromSitePayload(body, equipmentTemplates = []) {
  if (!body || typeof body !== 'object') {
    throw new HttpError(400, 'Invalid body');
  }
  const name = String(body.name || '').trim();
  if (!name) throw new HttpError(400, 'name is required');

  let appliesToEquipmentType = String(body.appliesToEquipmentType || '').trim();
  let globalEquipmentTemplateId = body.globalEquipmentTemplateId || null;
  let equipmentTemplateName = body.equipmentTemplateName != null ? String(body.equipmentTemplateName) : null;

  const siteEqId = body.equipmentTemplateId || null;
  const siteEq = siteEqId
    ? equipmentTemplates.find((e) => e && e.id === siteEqId)
    : equipmentTemplates.find(
        (e) =>
          e &&
          (e.name || '').toLowerCase() === String(body.appliesTo || '').toLowerCase()
      );

  if (!appliesToEquipmentType && siteEq) {
    appliesToEquipmentType = String(siteEq.equipmentType || 'CUSTOM').trim() || 'CUSTOM';
  }
  if (!appliesToEquipmentType) {
    appliesToEquipmentType = 'CUSTOM';
  }
  if (!equipmentTemplateName && siteEq) {
    equipmentTemplateName = siteEq.name || null;
  }

  const graphicEditorState =
    body.graphicEditorState != null && typeof body.graphicEditorState === 'object'
      ? body.graphicEditorState
      : null;
  const boundPointCount =
    typeof body.boundPointCount === 'number' && !Number.isNaN(body.boundPointCount)
      ? body.boundPointCount
      : countBindingsFromGraphicState(graphicEditorState);

  const created = await prisma.globalGraphicTemplate.create({
    data: {
      name,
      appliesToEquipmentType,
      globalEquipmentTemplateId: globalEquipmentTemplateId || null,
      equipmentTemplateName,
      graphicEditorStateJson: graphicEditorState,
      boundPointCount,
    },
  });

  return getGraphicTemplateById(created.id);
}

module.exports = {
  listEquipmentTemplates,
  getEquipmentTemplateById,
  createEquipmentTemplateFromSitePayload,
  updateEquipmentTemplateName,
  deleteEquipmentTemplate,
  listGraphicTemplates,
  getGraphicTemplateById,
  createGraphicTemplateFromSitePayload,
  updateGraphicTemplateName,
  deleteGraphicTemplate,
};

File: server/src/main/java/com/legioncontrols/server/templates/GlobalTemplateLibraryService.java
package com.legioncontrols.server.templates;

import com.legioncontrols.server.common.ApiException;
import com.legioncontrols.server.common.JsonUtil;
import com.legioncontrols.server.jooq.generated.enums.Entitystatus;
import com.legioncontrols.server.jooq.generated.tables.records.GlobalequipmenttemplateRecord;
import com.legioncontrols.server.jooq.generated.tables.records.GlobalgraphictemplateRecord;
import java.time.LocalDateTime;
import java.util.List;
import java.util.UUID;
import org.jooq.DSLContext;
import org.springframework.stereotype.Service;
import tools.jackson.databind.JsonNode;
import tools.jackson.databind.node.ArrayNode;

import static com.legioncontrols.server.jooq.generated.tables.Globalequipmenttemplate.GLOBALEQUIPMENTTEMPLATE;
import static com.legioncontrols.server.jooq.generated.tables.Globalgraphictemplate.GLOBALGRAPHICTEMPLATE;

/** Java equivalent of backend/src/modules/globalTemplateLibrary/globalTemplateLibrary.service.js. */
@Service
public class GlobalTemplateLibraryService {

    private final DSLContext dsl;
    private final JsonUtil json;
    private final LegionStarterEquipmentTemplates starterTemplates;
    private final Object seedLock = new Object();

    public GlobalTemplateLibraryService(DSLContext dsl, JsonUtil json, LegionStarterEquipmentTemplates starterTemplates) {
        this.dsl = dsl;
        this.json = json;
        this.starterTemplates = starterTemplates;
    }

    /** One-time style cleanup: legacy rows were named "Legion AHU", etc. */
    private void stripLegacyLegionPrefixFromGlobalEquipmentNames() {
        dsl.execute("UPDATE \"GlobalEquipmentTemplate\" SET name = TRIM(BOTH ' ' FROM REPLACE(name, 'Legion ', '')) WHERE name LIKE 'Legion %'");
    }

    /** If GlobalEquipmentTemplate has no rows, upsert starter templates. Avoids an empty library when the DB seed was never run. */
    private void ensureGlobalEquipmentStartersInDb() {
        synchronized (seedLock) {
            long count = dsl.selectCount().from(GLOBALEQUIPMENTTEMPLATE).fetchOne(0, Long.class);
            if (count > 0) return;
            LocalDateTime now = LocalDateTime.now();
            for (LegionStarterEquipmentTemplates.SeedRow row : starterTemplates.getGlobalStarterTemplateSeedRows()) {
                GlobalequipmenttemplateRecord existing = dsl.selectFrom(GLOBALEQUIPMENTTEMPLATE).where(GLOBALEQUIPMENTTEMPLATE.ID.eq(row.id())).fetchOne();
                GlobalequipmenttemplateRecord record = existing != null ? existing : dsl.newRecord(GLOBALEQUIPMENTTEMPLATE);
                if (existing == null) {
                    record.setId(row.id());
                    record.setCreatedat(now);
                }
                record.setName(row.name());
                record.setEquipmenttype(row.equipmentType());
                record.setDescription(row.description());
                record.setDefaultgraphicname(row.defaultGraphicName());
                record.setPointsjson(json.toJsonb(row.pointsJson()));
                record.setStatus(Entitystatus.ACTIVE);
                record.setUpdatedat(now);
                if (existing == null) record.insert(); else record.update();
            }
        }
    }

    private ArrayNode normalizePointsJson(JsonNode raw) {
        if (raw != null && raw.isArray()) return (ArrayNode) raw;
        return json.newArray();
    }

    public record EquipmentListRow(String id, String name, String equipmentType, int pointCount, String defaultGraphicName) {
    }

    private EquipmentListRow equipmentListRowFromRecord(GlobalequipmenttemplateRecord r) {
        ArrayNode points = normalizePointsJson(json.toJsonNode(r.getPointsjson()));
        return new EquipmentListRow(r.getId(), r.getName(), r.getEquipmenttype(), points.size(), r.getDefaultgraphicname());
    }

    public List<EquipmentListRow> listEquipmentTemplates() {
        stripLegacyLegionPrefixFromGlobalEquipmentNames();
        ensureGlobalEquipmentStartersInDb();
        return dsl.selectFrom(GLOBALEQUIPMENTTEMPLATE).where(GLOBALEQUIPMENTTEMPLATE.STATUS.eq(Entitystatus.ACTIVE))
            .orderBy(GLOBALEQUIPMENTTEMPLATE.NAME.asc()).fetch().stream().map(this::equipmentListRowFromRecord).toList();
    }

    public record EquipmentDetail(String id, String name, String equipmentType, String description,
                                   String defaultGraphicName, int pointCount, JsonNode points) {
    }

    public EquipmentDetail getEquipmentTemplateById(String id) {
        stripLegacyLegionPrefixFromGlobalEquipmentNames();
        ensureGlobalEquipmentStartersInDb();
        GlobalequipmenttemplateRecord r = dsl.selectFrom(GLOBALEQUIPMENTTEMPLATE)
            .where(GLOBALEQUIPMENTTEMPLATE.ID.eq(id), GLOBALEQUIPMENTTEMPLATE.STATUS.eq(Entitystatus.ACTIVE)).fetchOne();
        if (r == null) throw ApiException.notFound("Global equipment template not found");
        ArrayNode points = normalizePointsJson(json.toJsonNode(r.getPointsjson()));
        return new EquipmentDetail(r.getId(), r.getName(), r.getEquipmenttype(), r.getDescription() != null ? r.getDescription() : "",
            r.getDefaultgraphicname(), points.size(), points);
    }

    public EquipmentDetail createEquipmentTemplateFromSitePayload(JsonNode body) {
        if (body == null || !body.isObject()) throw ApiException.badRequest("Invalid body");
        String name = body.hasNonNull("name") ? body.get("name").asText().trim() : "";
        if (name.isEmpty()) throw ApiException.badRequest("name is required");
        String equipmentType = body.hasNonNull("equipmentType") && !body.get("equipmentType").asText().isBlank()
            ? body.get("equipmentType").asText().trim() : "CUSTOM";
        String description = body.hasNonNull("description") ? body.get("description").asText().trim() : "";
        String defaultGraphicName = firstNonBlank(body, "defaultGraphicName", "defaultGraphic");
        ArrayNode points = normalizePointsJson(body.get("points"));

        LocalDateTime now = LocalDateTime.now();
        GlobalequipmenttemplateRecord record = dsl.newRecord(GLOBALEQUIPMENTTEMPLATE);
        record.setId(UUID.randomUUID().toString());
        record.setName(name);
        record.setEquipmenttype(equipmentType);
        record.setDescription(description);
        record.setDefaultgraphicname(defaultGraphicName);
        record.setPointsjson(json.toJsonb(points));
        record.setCreatedat(now);
        record.setUpdatedat(now);
        record.insert();

        return getEquipmentTemplateById(record.getId());
    }

    private static String firstNonBlank(JsonNode body, String... fields) {
        for (String f : fields) {
            if (body.hasNonNull(f) && !body.get(f).asText().isBlank()) return body.get(f).asText().trim();
        }
        return null;
    }

    public EquipmentListRow updateEquipmentTemplateName(String id, JsonNode body) {
        String name = body != null && body.hasNonNull("name") ? body.get("name").asText().trim() : "";
        if (name.isEmpty()) throw ApiException.badRequest("name is required");
        GlobalequipmenttemplateRecord r = dsl.selectFrom(GLOBALEQUIPMENTTEMPLATE).where(GLOBALEQUIPMENTTEMPLATE.ID.eq(id)).fetchOne();
        if (r == null) throw ApiException.notFound("Global equipment template not found");
        r.setName(name);
        r.setUpdatedat(LocalDateTime.now());
        r.update();
        return equipmentListRowFromRecord(r);
    }

    public void deleteEquipmentTemplate(String id) {
        boolean exists = dsl.fetchExists(dsl.selectFrom(GLOBALEQUIPMENTTEMPLATE).where(GLOBALEQUIPMENTTEMPLATE.ID.eq(id)));
        if (!exists) throw ApiException.notFound("Global equipment template not found");
        dsl.deleteFrom(GLOBALEQUIPMENTTEMPLATE).where(GLOBALEQUIPMENTTEMPLATE.ID.eq(id)).execute();
    }

    private static int countBindingsFromGraphicState(JsonNode state) {
        if (state == null || !state.isObject()) return 0;
        JsonNode objects = state.get("objects");
        if (objects == null || !objects.isArray()) return 0;
        int count = 0;
        for (JsonNode o : objects) {
            if (o != null && o.hasNonNull("pointBinding") && !o.get("pointBinding").asText().isBlank()) count++;
        }
        return count;
    }

    public record GraphicListRow(String id, String name, String appliesToEquipmentType, int boundPointCount,
                                  String globalEquipmentTemplateId, String equipmentTemplateName) {
    }

    private GraphicListRow graphicListRowFromRecord(GlobalgraphictemplateRecord r) {
        return new GraphicListRow(r.getId(), r.getName(), r.getAppliestoequipmenttype(),
            r.getBoundpointcount() != null ? r.getBoundpointcount() : 0, r.getGlobalequipmenttemplateid(), r.getEquipmenttemplatename());
    }

    public GraphicListRow updateGraphicTemplateName(String id, JsonNode body) {
        String name = body != null && body.hasNonNull("name") ? body.get("name").asText().trim() : "";
        if (name.isEmpty()) throw ApiException.badRequest("name is required");
        GlobalgraphictemplateRecord r = dsl.selectFrom(GLOBALGRAPHICTEMPLATE).where(GLOBALGRAPHICTEMPLATE.ID.eq(id)).fetchOne();
        if (r == null) throw ApiException.notFound("Global graphic template not found");
        r.setName(name);
        r.setUpdatedat(LocalDateTime.now());
        r.update();
        return graphicListRowFromRecord(r);
    }

    public void deleteGraphicTemplate(String id) {
        boolean exists = dsl.fetchExists(dsl.selectFrom(GLOBALGRAPHICTEMPLATE).where(GLOBALGRAPHICTEMPLATE.ID.eq(id)));
        if (!exists) throw ApiException.notFound("Global graphic template not found");
        dsl.deleteFrom(GLOBALGRAPHICTEMPLATE).where(GLOBALGRAPHICTEMPLATE.ID.eq(id)).execute();
    }

    public List<GraphicListRow> listGraphicTemplates() {
        return dsl.selectFrom(GLOBALGRAPHICTEMPLATE).where(GLOBALGRAPHICTEMPLATE.STATUS.eq(Entitystatus.ACTIVE))
            .orderBy(GLOBALGRAPHICTEMPLATE.NAME.asc()).fetch().stream().map(this::graphicListRowFromRecord).toList();
    }

    public record GraphicDetail(String id, String name, String appliesToEquipmentType, String globalEquipmentTemplateId,
                                 String equipmentTemplateName, int boundPointCount, JsonNode graphicEditorState) {
    }

    public GraphicDetail getGraphicTemplateById(String id) {
        GlobalgraphictemplateRecord r = dsl.selectFrom(GLOBALGRAPHICTEMPLATE)
            .where(GLOBALGRAPHICTEMPLATE.ID.eq(id), GLOBALGRAPHICTEMPLATE.STATUS.eq(Entitystatus.ACTIVE)).fetchOne();
        if (r == null) throw ApiException.notFound("Global graphic template not found");
        JsonNode state = r.getGraphiceditorstatejson() != null ? json.toJsonNode(r.getGraphiceditorstatejson()) : null;
        state = state != null && state.isObject() ? state : null;
        int boundFromState = countBindingsFromGraphicState(state);
        int stored = r.getBoundpointcount() != null ? r.getBoundpointcount() : 0;
        return new GraphicDetail(r.getId(), r.getName(), r.getAppliestoequipmenttype(), r.getGlobalequipmenttemplateid(),
            r.getEquipmenttemplatename(), Math.max(stored, boundFromState), state);
    }

    public GraphicDetail createGraphicTemplateFromSitePayload(JsonNode body, JsonNode equipmentTemplates) {
        if (body == null || !body.isObject()) throw ApiException.badRequest("Invalid body");
        String name = body.hasNonNull("name") ? body.get("name").asText().trim() : "";
        if (name.isEmpty()) throw ApiException.badRequest("name is required");

        String appliesToEquipmentType = body.hasNonNull("appliesToEquipmentType") ? body.get("appliesToEquipmentType").asText().trim() : "";
        String globalEquipmentTemplateId = body.hasNonNull("globalEquipmentTemplateId") ? body.get("globalEquipmentTemplateId").asText() : null;
        String equipmentTemplateName = body.hasNonNull("equipmentTemplateName") ? body.get("equipmentTemplateName").asText() : null;

        JsonNode siteEq = null;
        String siteEqId = body.hasNonNull("equipmentTemplateId") ? body.get("equipmentTemplateId").asText() : null;
        if (equipmentTemplates != null && equipmentTemplates.isArray()) {
            for (JsonNode e : equipmentTemplates) {
                if (siteEqId != null) {
                    if (e.hasNonNull("id") && siteEqId.equals(e.get("id").asText())) { siteEq = e; break; }
                } else {
                    String appliesTo = body.hasNonNull("appliesTo") ? body.get("appliesTo").asText().toLowerCase() : "";
                    if (e.hasNonNull("name") && e.get("name").asText().toLowerCase().equals(appliesTo)) { siteEq = e; break; }
                }
            }
        }

        if (appliesToEquipmentType.isEmpty() && siteEq != null) {
            appliesToEquipmentType = siteEq.hasNonNull("equipmentType") && !siteEq.get("equipmentType").asText().isBlank()
                ? siteEq.get("equipmentType").asText().trim() : "CUSTOM";
        }
        if (appliesToEquipmentType.isEmpty()) appliesToEquipmentType = "CUSTOM";
        if (equipmentTemplateName == null && siteEq != null && siteEq.hasNonNull("name")) {
            equipmentTemplateName = siteEq.get("name").asText();
        }

        JsonNode graphicEditorState = body.get("graphicEditorState");
        graphicEditorState = graphicEditorState != null && graphicEditorState.isObject() ? graphicEditorState : null;
        int boundPointCount = body.has("boundPointCount") && body.get("boundPointCount").isNumber()
            ? body.get("boundPointCount").asInt() : countBindingsFromGraphicState(graphicEditorState);

        LocalDateTime now = LocalDateTime.now();
        GlobalgraphictemplateRecord record = dsl.newRecord(GLOBALGRAPHICTEMPLATE);
        record.setId(UUID.randomUUID().toString());
        record.setName(name);
        record.setAppliestoequipmenttype(appliesToEquipmentType);
        record.setGlobalequipmenttemplateid(globalEquipmentTemplateId);
        record.setEquipmenttemplatename(equipmentTemplateName);
        if (graphicEditorState != null) record.setGraphiceditorstatejson(json.toJsonb(graphicEditorState));
        record.setBoundpointcount(boundPointCount);
        record.setCreatedat(now);
        record.setUpdatedat(now);
        record.insert();

        return getGraphicTemplateById(record.getId());
    }
}
